cloudbridge/cloud/providers/azure/azure_client.py

```diff
@@ -286,6 +290,65 @@ def create_storage_account(self, name, params):
         return self.storage_client.storage_accounts. \
             create(self.resource_group, name.lower(), params).result()
 
+    # Create a storage account. To prevent a race condition, try
+    # to get or create at least twice
+    @tenacity.retry(stop=tenacity.stop_after_attempt(2),
+                    retry=tenacity.retry_if_exception_type(CloudError),
+                    reraise=True)
+    def _get_or_create_storage_account(self):
+        if self._storage_account:
+            return self._storage_account
+        else:
+            try:
+                self._storage_account = \
+                    self.get_storage_account(self.storage_account)
+            except CloudError as cloud_error:
+                if cloud_error.error.error == "ResourceNotFound":
+                    storage_account_params = {
+                        'sku': {
+                            'name': 'Standard_LRS'
+                        },
+                        'kind': 'storage',
+                        'location': self.region_name,
+                    }
+                    try:
+                        self._storage_account = \
+                            self.create_storage_account(self.storage_account,
+                                                        storage_account_params)
+                    except CloudError as cloud_error2:
+                        if cloud_error2.error.error == "AuthorizationFailed":
+                            mess = 'The following error was returned by ' \
+                                   'Azure:\n%s\n\nThis is likely because the' \
+                                   ' Role associated with the provided ' \
+                                   'credentials does not allow for Storage ' \
+                                   'Account creation.\nA Storage Account is ' \
+                                   'necessary in order to perform the ' \
+                                   'desired operation. You must either ' \
+                                   'provide an existing Storage Account name' \
+                                   ' as part of the configuration, or ' \
+                                   'elevate the associated Role.\nFor more ' \
```

Comment:
I would propose removing the advice to elevate the associated Role, since we don't really endorse that? We could just suggest that they should provide a SA name as part of the cloudbridge config?

Reply:
Since what we are trying to achieve with cloudbridge is to make it easy for someone to write code that works on multiple clouds, without having to do much testing etc., someone who is not that familiar with azure should be naturally guided towards the right approach, whatever that might be. My understanding was that the right approach here was to specify the storage account name. Is that not the case?

I think it's ok to leave the instructions on elevating privileges etc. as long as we point out what the "proper"/recommended approach is. Right now, I'm not really sure how role-wise, the creation of the storage account is separated from the usage of a storage account, i.e. who does what from a practical standpoint. Does this mean that some higher privilege user creates the storage account for you, and then, at launch time, a cloudlaunch user simply specifies the SA name? Does that mean we need a GUI element in cloudlaunch to specify the SA name?

Reply:
> My understanding was that the right approach here was to specify the storage account name. Is that not the case?

There's not a right and wrong way here. For simplicity sake of creating a provider config, we suggest a user specifies only the 4 required fields and a SA will be reused or created. However, creating it requires _owner_ or _contributor_ level role. A user may or may not have that privilege.

> Does this mean that some higher privilege user creates the storage account for you, and then, at launch time, a cloudlaunch user simply specifies the SA name? 

Possibly. If a user doesn’t have the permissions, they'll have to supply a SA name; if they don't supply it, we/CloudBridge will try to create it and possibly fail with the above message indicating to the user they have two two options.

> Does that mean we need a GUI element in cloudlaunch to specify the SA name?

Yes, just a another (optional) text field in the provider credentials form.

In general, I feel an elaboration of all of this belongs in the docs, not a comment in the exception. There, we can have a more explanatory text with the recommended approach while this here is just an informative error message that ideally gives an unsuspecting user a path forward.